{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7e46ded5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>id</th>\n",
       "      <th>parents</th>\n",
       "      <th>definition</th>\n",
       "      <th>aspect</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GO:0044208</th>\n",
       "      <td>'de novo' AMP biosynthetic process</td>\n",
       "      <td>GO:0044208</td>\n",
       "      <td>GO:0006167</td>\n",
       "      <td>The chemical reactions and pathways resulting ...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0044210</th>\n",
       "      <td>'de novo' CTP biosynthetic process</td>\n",
       "      <td>GO:0044210</td>\n",
       "      <td>GO:0006241</td>\n",
       "      <td>The chemical reactions and pathways resulting ...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042351</th>\n",
       "      <td>'de novo' GDP-L-fucose biosynthetic process</td>\n",
       "      <td>GO:0042351</td>\n",
       "      <td>GO:0034654|GO:0046368|GO:1901137</td>\n",
       "      <td>The chemical reactions and pathways resulting ...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006189</th>\n",
       "      <td>'de novo' IMP biosynthetic process</td>\n",
       "      <td>GO:0006189</td>\n",
       "      <td>GO:0006188</td>\n",
       "      <td>The chemical reactions and pathways resulting ...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0034627</th>\n",
       "      <td>'de novo' NAD biosynthetic process</td>\n",
       "      <td>GO:0034627</td>\n",
       "      <td>GO:0009435</td>\n",
       "      <td>The chemical reactions and pathways resulting ...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0035375</th>\n",
       "      <td>zymogen binding</td>\n",
       "      <td>GO:0035375</td>\n",
       "      <td>GO:0019899</td>\n",
       "      <td>Interacting selectively and non-covalently wit...</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042588</th>\n",
       "      <td>zymogen granule</td>\n",
       "      <td>GO:0042588</td>\n",
       "      <td>GO:0030141</td>\n",
       "      <td>A membrane-bounded, cytoplasmic secretory gran...</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0070625</th>\n",
       "      <td>zymogen granule exocytosis</td>\n",
       "      <td>GO:0070625</td>\n",
       "      <td>GO:0017156</td>\n",
       "      <td>The release of intracellular molecules contain...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042589</th>\n",
       "      <td>zymogen granule membrane</td>\n",
       "      <td>GO:0042589</td>\n",
       "      <td>GO:0030667</td>\n",
       "      <td>The lipid bilayer surrounding a zymogen granule.</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0097341</th>\n",
       "      <td>zymogen inhibition</td>\n",
       "      <td>GO:0097341</td>\n",
       "      <td>GO:0010955</td>\n",
       "      <td>Any process that prevents the proteolytic proc...</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18836 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  label          id  \\\n",
       "id                                                                    \n",
       "GO:0044208           'de novo' AMP biosynthetic process  GO:0044208   \n",
       "GO:0044210           'de novo' CTP biosynthetic process  GO:0044210   \n",
       "GO:0042351  'de novo' GDP-L-fucose biosynthetic process  GO:0042351   \n",
       "GO:0006189           'de novo' IMP biosynthetic process  GO:0006189   \n",
       "GO:0034627           'de novo' NAD biosynthetic process  GO:0034627   \n",
       "...                                                 ...         ...   \n",
       "GO:0035375                              zymogen binding  GO:0035375   \n",
       "GO:0042588                              zymogen granule  GO:0042588   \n",
       "GO:0070625                   zymogen granule exocytosis  GO:0070625   \n",
       "GO:0042589                     zymogen granule membrane  GO:0042589   \n",
       "GO:0097341                           zymogen inhibition  GO:0097341   \n",
       "\n",
       "                                     parents  \\\n",
       "id                                             \n",
       "GO:0044208                        GO:0006167   \n",
       "GO:0044210                        GO:0006241   \n",
       "GO:0042351  GO:0034654|GO:0046368|GO:1901137   \n",
       "GO:0006189                        GO:0006188   \n",
       "GO:0034627                        GO:0009435   \n",
       "...                                      ...   \n",
       "GO:0035375                        GO:0019899   \n",
       "GO:0042588                        GO:0030141   \n",
       "GO:0070625                        GO:0017156   \n",
       "GO:0042589                        GO:0030667   \n",
       "GO:0097341                        GO:0010955   \n",
       "\n",
       "                                                   definition  \\\n",
       "id                                                              \n",
       "GO:0044208  The chemical reactions and pathways resulting ...   \n",
       "GO:0044210  The chemical reactions and pathways resulting ...   \n",
       "GO:0042351  The chemical reactions and pathways resulting ...   \n",
       "GO:0006189  The chemical reactions and pathways resulting ...   \n",
       "GO:0034627  The chemical reactions and pathways resulting ...   \n",
       "...                                                       ...   \n",
       "GO:0035375  Interacting selectively and non-covalently wit...   \n",
       "GO:0042588  A membrane-bounded, cytoplasmic secretory gran...   \n",
       "GO:0070625  The release of intracellular molecules contain...   \n",
       "GO:0042589   The lipid bilayer surrounding a zymogen granule.   \n",
       "GO:0097341  Any process that prevents the proteolytic proc...   \n",
       "\n",
       "                        aspect  \n",
       "id                              \n",
       "GO:0044208  biological_process  \n",
       "GO:0044210  biological_process  \n",
       "GO:0042351  biological_process  \n",
       "GO:0006189  biological_process  \n",
       "GO:0034627  biological_process  \n",
       "...                        ...  \n",
       "GO:0035375  molecular_function  \n",
       "GO:0042588  cellular_component  \n",
       "GO:0070625  biological_process  \n",
       "GO:0042589  cellular_component  \n",
       "GO:0097341  biological_process  \n",
       "\n",
       "[18836 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import networkx as nx\n",
    "from annoq_tree_gen import DTYPES\n",
    "\n",
    "annotations = pd.read_csv(\"./../../annoq-data/tree.csv\", sep=',', dtype=DTYPES)\n",
    "terms = pd.read_csv(\"./../../annoq-data/slim/terms.tsv\", sep='\\t')\n",
    "\n",
    "annotations.set_index(\"id\", inplace=True, drop=False)\n",
    "\n",
    "term_cols = {\n",
    "    'ID':'id',\n",
    "    'LABEL':'label',\n",
    "    'SubClass Of':'parents',\n",
    "    'hasOBONamespace':'aspect'\n",
    "}\n",
    "\n",
    "terms.rename(columns=term_cols, inplace=True)\n",
    "terms.set_index(\"id\", inplace=True, drop=False)\n",
    "terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4bd5d8cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>leaf</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>detail</th>\n",
       "      <th>link</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sort</th>\n",
       "      <th>value_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>root</td>\n",
       "      <td>Annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>Basic Info</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basic information about the variant, such as c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ANNOVAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pre-computed ANNOVAR annotations for all alter...</td>\n",
       "      <td>http://annovar.openbioinformatics.org/en/lates...</td>\n",
       "      <td>20601685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>208</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>SnpEff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AnpEff is a program for annotating and predict...</td>\n",
       "      <td>http://pcingola.github.io/SnpEff/</td>\n",
       "      <td>22728672</td>\n",
       "      <td>3.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>132</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>VEP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant Effect Predictor (VEP) is developed by...</td>\n",
       "      <td>https://uswest.ensembl.org/info/docs/tools/vep...</td>\n",
       "      <td>27268795</td>\n",
       "      <td>4.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>621</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>sno_miRNA_type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>the type of snoRNA or miRNA (from miRBase/snoR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>622</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_ada_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>623</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_rf_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>624</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>target_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target gene (for promoter, enhancer, etc.) bas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>625</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>ultra_sensitive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>whether defined as ultra-sensitive region base...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>626 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id parent_id   leaf                          name       label  \\\n",
       "id                                                                    \n",
       "0      0       NaN  False                          root  Annotation   \n",
       "1      1         0  False                    Basic Info         NaN   \n",
       "26    26         0  False                       ANNOVAR         NaN   \n",
       "208  208         0  False                        SnpEff         NaN   \n",
       "132  132         0  False                           VEP         NaN   \n",
       "..   ...       ...    ...                           ...         ...   \n",
       "621  621       495   True                sno_miRNA_type         NaN   \n",
       "622  622       495   True  splicing_consensus_ada_score         NaN   \n",
       "623  623       495   True   splicing_consensus_rf_score         NaN   \n",
       "624  624       495   True                   target_gene         NaN   \n",
       "625  625       495   True               ultra_sensitive         NaN   \n",
       "\n",
       "                                                detail  \\\n",
       "id                                                       \n",
       "0                                                  NaN   \n",
       "1    Basic information about the variant, such as c...   \n",
       "26   Pre-computed ANNOVAR annotations for all alter...   \n",
       "208  AnpEff is a program for annotating and predict...   \n",
       "132  Variant Effect Predictor (VEP) is developed by...   \n",
       "..                                                 ...   \n",
       "621  the type of snoRNA or miRNA (from miRBase/snoR...   \n",
       "622  splicing-change prediction for splicing consen...   \n",
       "623  splicing-change prediction for splicing consen...   \n",
       "624  target gene (for promoter, enhancer, etc.) bas...   \n",
       "625  whether defined as ultra-sensitive region base...   \n",
       "\n",
       "                                                  link      pmid  sort  \\\n",
       "id                                                                       \n",
       "0                                                  NaN       NaN   0.0   \n",
       "1                                                  NaN       NaN   1.0   \n",
       "26   http://annovar.openbioinformatics.org/en/lates...  20601685   2.0   \n",
       "208                  http://pcingola.github.io/SnpEff/  22728672   3.0   \n",
       "132  https://uswest.ensembl.org/info/docs/tools/vep...  27268795   4.0   \n",
       "..                                                 ...       ...   ...   \n",
       "621                                                NaN       NaN   NaN   \n",
       "622                                                NaN       NaN   NaN   \n",
       "623                                                NaN       NaN   NaN   \n",
       "624                                                NaN       NaN   NaN   \n",
       "625                                                NaN       NaN   NaN   \n",
       "\n",
       "    value_type  \n",
       "id              \n",
       "0         None  \n",
       "1         None  \n",
       "26        None  \n",
       "208       None  \n",
       "132       None  \n",
       "..         ...  \n",
       "621       None  \n",
       "622       None  \n",
       "623       None  \n",
       "624       None  \n",
       "625       None  \n",
       "\n",
       "[626 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def is_leaf(G, node):\n",
    "    return G.out_degree(node) == 0 and G.in_degree(node) == 1\n",
    "\n",
    "def add_value_type(name):\n",
    "    if 'GO_' in name and 'list_id' in name:\n",
    "        return 'GO_id'\n",
    "    if 'GO_' in name and 'list' in name:\n",
    "        return 'GO_label'\n",
    "        \n",
    "    return None\n",
    "\n",
    "\n",
    "annotations['value_type'] = annotations.apply(lambda row: add_value_type(row['name']),axis=1)\n",
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ebc45572",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g = nx.from_pandas_edgelist(\n",
    "    annotations, 'parent_id', 'id', create_using=nx.DiGraph())\n",
    "\n",
    "roots = [annotations.loc[n]['name'] for n in nx.descendants(g,'26') if is_leaf(g, n)]\n",
    "\n",
    "#len(roots)\n",
    "\n",
    "def leaf_count(G, source):\n",
    "    children = [n for n in nx.descendants(G, source) if is_leaf(g, n)]\n",
    "    return len(children)\n",
    "\n",
    "\n",
    "def descendants_count(G, source):\n",
    "    children = [n for n in nx.descendants(G, source)]\n",
    "    return len(children)\n",
    "\n",
    "def parents_count(G, source):\n",
    "    children = [n for n in nx.descendants(G, source) if not is_leaf(g, n)]\n",
    "    return len(children)\n",
    "\n",
    "child_count(g, '26')\n",
    "#nx.descendants(g,'26')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ddc1a65e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>leaf</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>detail</th>\n",
       "      <th>link</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sort</th>\n",
       "      <th>value_type</th>\n",
       "      <th>count</th>\n",
       "      <th>children_count</th>\n",
       "      <th>category_count</th>\n",
       "      <th>leaf_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>root</td>\n",
       "      <td>Annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>607</td>\n",
       "      <td>625</td>\n",
       "      <td>18</td>\n",
       "      <td>607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>Basic Info</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basic information about the variant, such as c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ANNOVAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pre-computed ANNOVAR annotations for all alter...</td>\n",
       "      <td>http://annovar.openbioinformatics.org/en/lates...</td>\n",
       "      <td>20601685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>208</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>SnpEff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AnpEff is a program for annotating and predict...</td>\n",
       "      <td>http://pcingola.github.io/SnpEff/</td>\n",
       "      <td>22728672</td>\n",
       "      <td>3.0</td>\n",
       "      <td>None</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>132</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>VEP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant Effect Predictor (VEP) is developed by...</td>\n",
       "      <td>https://uswest.ensembl.org/info/docs/tools/vep...</td>\n",
       "      <td>27268795</td>\n",
       "      <td>4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>621</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>sno_miRNA_type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>the type of snoRNA or miRNA (from miRBase/snoR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>622</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_ada_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>623</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_rf_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>624</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>target_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target gene (for promoter, enhancer, etc.) bas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>625</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>ultra_sensitive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>whether defined as ultra-sensitive region base...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>626 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id parent_id   leaf                          name       label  \\\n",
       "id                                                                    \n",
       "0      0       NaN  False                          root  Annotation   \n",
       "1      1         0  False                    Basic Info         NaN   \n",
       "26    26         0  False                       ANNOVAR         NaN   \n",
       "208  208         0  False                        SnpEff         NaN   \n",
       "132  132         0  False                           VEP         NaN   \n",
       "..   ...       ...    ...                           ...         ...   \n",
       "621  621       495   True                sno_miRNA_type         NaN   \n",
       "622  622       495   True  splicing_consensus_ada_score         NaN   \n",
       "623  623       495   True   splicing_consensus_rf_score         NaN   \n",
       "624  624       495   True                   target_gene         NaN   \n",
       "625  625       495   True               ultra_sensitive         NaN   \n",
       "\n",
       "                                                detail  \\\n",
       "id                                                       \n",
       "0                                                  NaN   \n",
       "1    Basic information about the variant, such as c...   \n",
       "26   Pre-computed ANNOVAR annotations for all alter...   \n",
       "208  AnpEff is a program for annotating and predict...   \n",
       "132  Variant Effect Predictor (VEP) is developed by...   \n",
       "..                                                 ...   \n",
       "621  the type of snoRNA or miRNA (from miRBase/snoR...   \n",
       "622  splicing-change prediction for splicing consen...   \n",
       "623  splicing-change prediction for splicing consen...   \n",
       "624  target gene (for promoter, enhancer, etc.) bas...   \n",
       "625  whether defined as ultra-sensitive region base...   \n",
       "\n",
       "                                                  link      pmid  sort  \\\n",
       "id                                                                       \n",
       "0                                                  NaN       NaN   0.0   \n",
       "1                                                  NaN       NaN   1.0   \n",
       "26   http://annovar.openbioinformatics.org/en/lates...  20601685   2.0   \n",
       "208                  http://pcingola.github.io/SnpEff/  22728672   3.0   \n",
       "132  https://uswest.ensembl.org/info/docs/tools/vep...  27268795   4.0   \n",
       "..                                                 ...       ...   ...   \n",
       "621                                                NaN       NaN   NaN   \n",
       "622                                                NaN       NaN   NaN   \n",
       "623                                                NaN       NaN   NaN   \n",
       "624                                                NaN       NaN   NaN   \n",
       "625                                                NaN       NaN   NaN   \n",
       "\n",
       "    value_type  count  children_count  category_count  leaf_count  \n",
       "id                                                                 \n",
       "0         None    607             625              18         607  \n",
       "1         None      5               5               0           5  \n",
       "26        None     56              56               0          56  \n",
       "208       None     79              79               0          79  \n",
       "132       None     75              75               0          75  \n",
       "..         ...    ...             ...             ...         ...  \n",
       "621       None      0               0               0           0  \n",
       "622       None      0               0               0           0  \n",
       "623       None      0               0               0           0  \n",
       "624       None      0               0               0           0  \n",
       "625       None      0               0               0           0  \n",
       "\n",
       "[626 rows x 14 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations['children_count'] = annotations.apply(lambda row: descendants_count(g, row['id']),axis=1)\n",
    "annotations['category_count'] = annotations.apply(lambda row: parents_count(g, row['id']),axis=1)\n",
    "annotations['leaf_count'] = annotations.apply(lambda row: leaf_count(g, row['id']),axis=1)\n",
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "29d8ec24",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations.to_csv('./../../annoq-data/tree-with-count.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
