{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7e46ded5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import networkx as nx\n",
    "from annoq_tree_gen import DTYPES\n",
    "from base import load_json\n",
    "\n",
    "annotations = pd.read_csv(\"./../../annoq-data/revised_tree.csv\", sep=',', dtype=DTYPES)\n",
    "terms = pd.read_csv(\"./../../annoq-data/slim/terms.tsv\", sep='\\t')\n",
    "mappings = load_json(\"./../../annoq-data/mappings.json\")\n",
    "\n",
    "annotations.set_index(\"id\", inplace=True, drop=False)\n",
    "\n",
    "term_cols = {\n",
    "    'ID':'id',\n",
    "    'LABEL':'label',\n",
    "    'SubClass Of':'parents',\n",
    "    'hasOBONamespace':'aspect'\n",
    "}\n",
    "\n",
    "terms.rename(columns=term_cols, inplace=True)\n",
    "terms.set_index(\"id\", inplace=True, drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4bd5d8cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>leaf</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>detail</th>\n",
       "      <th>link</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sort</th>\n",
       "      <th>value_type</th>\n",
       "      <th>count</th>\n",
       "      <th>children_count</th>\n",
       "      <th>category_count</th>\n",
       "      <th>leaf_count</th>\n",
       "      <th>root URL</th>\n",
       "      <th>sample URL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>root</td>\n",
       "      <td>Annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>607</td>\n",
       "      <td>625</td>\n",
       "      <td>18</td>\n",
       "      <td>607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>Basic Info</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basic information about the variant, such as c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ANNOVAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pre-computed ANNOVAR annotations for all alter...</td>\n",
       "      <td>http://annovar.openbioinformatics.org/en/lates...</td>\n",
       "      <td>20601685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>208</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>SnpEff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AnpEff is a program for annotating and predict...</td>\n",
       "      <td>http://pcingola.github.io/SnpEff/</td>\n",
       "      <td>22728672</td>\n",
       "      <td>3.0</td>\n",
       "      <td>None</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>132</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>VEP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant Effect Predictor (VEP) is developed by...</td>\n",
       "      <td>https://uswest.ensembl.org/info/docs/tools/vep...</td>\n",
       "      <td>27268795</td>\n",
       "      <td>4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>621</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>sno_miRNA_type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>the type of snoRNA or miRNA (from miRBase/snoR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>622</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_ada_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>623</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_rf_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>624</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>target_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target gene (for promoter, enhancer, etc.) bas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>625</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>ultra_sensitive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>whether defined as ultra-sensitive region base...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>626 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id parent_id   leaf                          name       label  \\\n",
       "id                                                                    \n",
       "0      0       NaN  False                          root  Annotation   \n",
       "1      1         0  False                    Basic Info         NaN   \n",
       "26    26         0  False                       ANNOVAR         NaN   \n",
       "208  208         0  False                        SnpEff         NaN   \n",
       "132  132         0  False                           VEP         NaN   \n",
       "..   ...       ...    ...                           ...         ...   \n",
       "621  621       495   True                sno_miRNA_type         NaN   \n",
       "622  622       495   True  splicing_consensus_ada_score         NaN   \n",
       "623  623       495   True   splicing_consensus_rf_score         NaN   \n",
       "624  624       495   True                   target_gene         NaN   \n",
       "625  625       495   True               ultra_sensitive         NaN   \n",
       "\n",
       "                                                detail  \\\n",
       "id                                                       \n",
       "0                                                  NaN   \n",
       "1    Basic information about the variant, such as c...   \n",
       "26   Pre-computed ANNOVAR annotations for all alter...   \n",
       "208  AnpEff is a program for annotating and predict...   \n",
       "132  Variant Effect Predictor (VEP) is developed by...   \n",
       "..                                                 ...   \n",
       "621  the type of snoRNA or miRNA (from miRBase/snoR...   \n",
       "622  splicing-change prediction for splicing consen...   \n",
       "623  splicing-change prediction for splicing consen...   \n",
       "624  target gene (for promoter, enhancer, etc.) bas...   \n",
       "625  whether defined as ultra-sensitive region base...   \n",
       "\n",
       "                                                  link      pmid  sort  \\\n",
       "id                                                                       \n",
       "0                                                  NaN       NaN   0.0   \n",
       "1                                                  NaN       NaN   1.0   \n",
       "26   http://annovar.openbioinformatics.org/en/lates...  20601685   2.0   \n",
       "208                  http://pcingola.github.io/SnpEff/  22728672   3.0   \n",
       "132  https://uswest.ensembl.org/info/docs/tools/vep...  27268795   4.0   \n",
       "..                                                 ...       ...   ...   \n",
       "621                                                NaN       NaN   NaN   \n",
       "622                                                NaN       NaN   NaN   \n",
       "623                                                NaN       NaN   NaN   \n",
       "624                                                NaN       NaN   NaN   \n",
       "625                                                NaN       NaN   NaN   \n",
       "\n",
       "    value_type  count  children_count  category_count  leaf_count root URL  \\\n",
       "id                                                                           \n",
       "0         None    607             625              18         607      NaN   \n",
       "1         None      5               5               0           5      NaN   \n",
       "26        None     56              56               0          56      NaN   \n",
       "208       None     79              79               0          79      NaN   \n",
       "132       None     75              75               0          75      NaN   \n",
       "..         ...    ...             ...             ...         ...      ...   \n",
       "621       None      0               0               0           0      NaN   \n",
       "622       None      0               0               0           0      NaN   \n",
       "623       None      0               0               0           0      NaN   \n",
       "624       None      0               0               0           0      NaN   \n",
       "625       None      0               0               0           0      NaN   \n",
       "\n",
       "    sample URL  \n",
       "id              \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "26         NaN  \n",
       "208        NaN  \n",
       "132        NaN  \n",
       "..         ...  \n",
       "621        NaN  \n",
       "622        NaN  \n",
       "623        NaN  \n",
       "624        NaN  \n",
       "625        NaN  \n",
       "\n",
       "[626 rows x 16 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def is_leaf(G, node):\n",
    "    return G.out_degree(node) == 0 and G.in_degree(node) == 1\n",
    "\n",
    "def add_value_type(name):\n",
    "    if 'GO_' in name and 'list_id' in name:\n",
    "        return 'GO_id'\n",
    "    if 'GO_' in name and 'list' in name:\n",
    "        return 'GO_label'\n",
    "        \n",
    "    return None\n",
    "\n",
    "\n",
    "annotations['value_type'] = annotations.apply(lambda row: add_value_type(row['name']),axis=1)\n",
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ebc45572",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'text'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g = nx.from_pandas_edgelist(\n",
    "    annotations, 'parent_id', 'id', create_using=nx.DiGraph())\n",
    "\n",
    "roots = [annotations.loc[n]['name'] for n in nx.descendants(g,'26') if is_leaf(g, n)]\n",
    "\n",
    "#len(roots)\n",
    "\n",
    "def leaf_count(G, source):\n",
    "    children = [n for n in nx.descendants(G, source) if is_leaf(g, n)]\n",
    "    return len(children)\n",
    "\n",
    "\n",
    "def descendants_count(G, source):\n",
    "    children = [n for n in nx.descendants(G, source)]\n",
    "    return len(children)\n",
    "\n",
    "def parents_count(G, source):\n",
    "    children = [n for n in nx.descendants(G, source) if not is_leaf(g, n)]\n",
    "    return len(children)\n",
    "\n",
    "def get_type(mapping_dict, field):\n",
    "    try:\n",
    "        field_key =  mapping_dict[\"annoq-test\"][\"mappings\"][\"properties\"][field]\n",
    "\n",
    "        if field_key != None:\n",
    "            return field_key['type']\n",
    "    \n",
    "    except KeyError:\n",
    "        return np.nan\n",
    "        \n",
    "    \n",
    "get_type(mappings, 'ref')\n",
    "#child_count(g, '26')\n",
    "#nx.descendants(g,'26')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ddc1a65e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>leaf</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "      <th>detail</th>\n",
       "      <th>link</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sort</th>\n",
       "      <th>value_type</th>\n",
       "      <th>count</th>\n",
       "      <th>children_count</th>\n",
       "      <th>category_count</th>\n",
       "      <th>leaf_count</th>\n",
       "      <th>root URL</th>\n",
       "      <th>sample URL</th>\n",
       "      <th>field_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>root</td>\n",
       "      <td>Annotation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>None</td>\n",
       "      <td>607</td>\n",
       "      <td>625</td>\n",
       "      <td>18</td>\n",
       "      <td>607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>Basic Info</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basic information about the variant, such as c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>ANNOVAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pre-computed ANNOVAR annotations for all alter...</td>\n",
       "      <td>http://annovar.openbioinformatics.org/en/lates...</td>\n",
       "      <td>20601685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>208</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>SnpEff</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AnpEff is a program for annotating and predict...</td>\n",
       "      <td>http://pcingola.github.io/SnpEff/</td>\n",
       "      <td>22728672</td>\n",
       "      <td>3.0</td>\n",
       "      <td>None</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>132</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>VEP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Variant Effect Predictor (VEP) is developed by...</td>\n",
       "      <td>https://uswest.ensembl.org/info/docs/tools/vep...</td>\n",
       "      <td>27268795</td>\n",
       "      <td>4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>621</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>sno_miRNA_type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>the type of snoRNA or miRNA (from miRBase/snoR...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>text</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>622</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_ada_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>623</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>splicing_consensus_rf_score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>splicing-change prediction for splicing consen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>624</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>target_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target gene (for promoter, enhancer, etc.) bas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>text</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>625</td>\n",
       "      <td>495</td>\n",
       "      <td>True</td>\n",
       "      <td>ultra_sensitive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>whether defined as ultra-sensitive region base...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>text</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>626 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id parent_id   leaf                          name       label  \\\n",
       "id                                                                    \n",
       "0      0       NaN  False                          root  Annotation   \n",
       "1      1         0  False                    Basic Info         NaN   \n",
       "26    26         0  False                       ANNOVAR         NaN   \n",
       "208  208         0  False                        SnpEff         NaN   \n",
       "132  132         0  False                           VEP         NaN   \n",
       "..   ...       ...    ...                           ...         ...   \n",
       "621  621       495   True                sno_miRNA_type         NaN   \n",
       "622  622       495   True  splicing_consensus_ada_score         NaN   \n",
       "623  623       495   True   splicing_consensus_rf_score         NaN   \n",
       "624  624       495   True                   target_gene         NaN   \n",
       "625  625       495   True               ultra_sensitive         NaN   \n",
       "\n",
       "                                                detail  \\\n",
       "id                                                       \n",
       "0                                                  NaN   \n",
       "1    Basic information about the variant, such as c...   \n",
       "26   Pre-computed ANNOVAR annotations for all alter...   \n",
       "208  AnpEff is a program for annotating and predict...   \n",
       "132  Variant Effect Predictor (VEP) is developed by...   \n",
       "..                                                 ...   \n",
       "621  the type of snoRNA or miRNA (from miRBase/snoR...   \n",
       "622  splicing-change prediction for splicing consen...   \n",
       "623  splicing-change prediction for splicing consen...   \n",
       "624  target gene (for promoter, enhancer, etc.) bas...   \n",
       "625  whether defined as ultra-sensitive region base...   \n",
       "\n",
       "                                                  link      pmid  sort  \\\n",
       "id                                                                       \n",
       "0                                                  NaN       NaN   0.0   \n",
       "1                                                  NaN       NaN   1.0   \n",
       "26   http://annovar.openbioinformatics.org/en/lates...  20601685   2.0   \n",
       "208                  http://pcingola.github.io/SnpEff/  22728672   3.0   \n",
       "132  https://uswest.ensembl.org/info/docs/tools/vep...  27268795   4.0   \n",
       "..                                                 ...       ...   ...   \n",
       "621                                                NaN       NaN   NaN   \n",
       "622                                                NaN       NaN   NaN   \n",
       "623                                                NaN       NaN   NaN   \n",
       "624                                                NaN       NaN   NaN   \n",
       "625                                                NaN       NaN   NaN   \n",
       "\n",
       "    value_type  count  children_count  category_count  leaf_count root URL  \\\n",
       "id                                                                           \n",
       "0         None    607             625              18         607      NaN   \n",
       "1         None      5               5               0           5      NaN   \n",
       "26        None     56              56               0          56      NaN   \n",
       "208       None     79              79               0          79      NaN   \n",
       "132       None     75              75               0          75      NaN   \n",
       "..         ...    ...             ...             ...         ...      ...   \n",
       "621       None      0               0               0           0      NaN   \n",
       "622       None      0               0               0           0      NaN   \n",
       "623       None      0               0               0           0      NaN   \n",
       "624       None      0               0               0           0      NaN   \n",
       "625       None      0               0               0           0      NaN   \n",
       "\n",
       "    sample URL field_type  \n",
       "id                         \n",
       "0          NaN        NaN  \n",
       "1          NaN        NaN  \n",
       "26         NaN        NaN  \n",
       "208        NaN        NaN  \n",
       "132        NaN        NaN  \n",
       "..         ...        ...  \n",
       "621        NaN       text  \n",
       "622        NaN      float  \n",
       "623        NaN      float  \n",
       "624        NaN       text  \n",
       "625        NaN       text  \n",
       "\n",
       "[626 rows x 17 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotations['children_count'] = annotations.apply(lambda row: descendants_count(g, row['id']),axis=1)\n",
    "annotations['category_count'] = annotations.apply(lambda row: parents_count(g, row['id']),axis=1)\n",
    "annotations['leaf_count'] = annotations.apply(lambda row: leaf_count(g, row['id']),axis=1)\n",
    "annotations['field_type'] = annotations.apply(lambda row: get_type(mappings, row['name']),axis=1)\n",
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "29d8ec24",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations.to_csv('./../../annoq-data/tree-with-types.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
